modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,385 @@
+    Ok(Response::new().add_attribute("action", "check_in"))
+}
+
+fn cast_vote(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    _app: &ChallengeApp,
+    vote: Option<bool>,
+    challenge_id: &str,
+) -> AppResult {
+    let today = date_from_block(&env);
+
+    // If Admin checked in today, friends can't vote
+    if DAILY_CHECK_INS
+        .load(deps.storage, today.clone())
+        .map_or(false, |check_in| {
+            check_in.last_checked_in.as_deref() == Some(&today)
+        })
+    {
+        return Err(AppError::AlreadyCheckedIn {});
+    }
+
+    // If the vote is None, default to true (meaning we assume Admin fulfilled his challenge)
+    let final_vote = vote.unwrap_or(true);
+
+    // Construct the vote
+    let vote_entry = Vote {
+        voter: info.sender.to_string(),
+        vote: Some(final_vote),
+        challenge_id: challenge_id.to_owned(),
+    };
+
+    // Load existing votes for the current block height or initialize an empty list if none exist
+    let mut votes_for_block = VOTES
+        .load(deps.storage, challenge_id.to_owned())
+        .unwrap_or_else(|_| Vec::new());
+
+    // Append the new vote
+    votes_for_block.push(vote_entry);
+
+    // Save the updated votes
+    VOTES.save(deps.storage, challenge_id.to_owned(), &votes_for_block)?;
+
+    Ok(Response::new().add_attribute("action", "cast_vote"))
+}
+
+fn count_votes(
+    deps: DepsMut,
+    info: MessageInfo,
+    _env: Env,
+    app: &ChallengeApp,
+    challenge_id: String,
+) -> AppResult {
+    let votes_for_challenge = VOTES
+        .load(deps.storage, challenge_id.clone())
+        .unwrap_or_else(|_| Vec::new());
+
+    let any_false_vote = votes_for_challenge.iter().any(|v| v.vote == Some(false));
+    if any_false_vote {
+        return charge_penalty(deps, info, app, challenge_id);
+    }
+
+    Ok(Response::new().add_attribute("action", "count_votes"))
+}
+
+fn charge_penalty(
+    deps: DepsMut,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: String,
+) -> AppResult {
+    let challenge = CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    let friends = CHALLENGE_FRIENDS.load(deps.storage, challenge_id.clone())?;
+
+    let admin_address = deps.api.addr_validate(info.sender.as_ref())?;
+    let bank = app.bank(deps.as_ref());
+    let executor = app.executor(deps.as_ref());
+
+    match challenge.collateral {
+        Penalty::FixedAmount { asset } => {
+            let num_friends = friends.len() as u128;
+            if num_friends == 0 {
+                return Err(AppError::Std(StdError::generic_err(
+                    "No friends found for the challenge.",
+                )));
+            }
+
+            // Calculate each friend's share
+            let amount_per_friend = asset.amount.u128() / num_friends;
+            let asset_per_friend = OfferAsset {
+                name: asset.name,
+                amount: Uint128::from(amount_per_friend),
+            };
```

Comment:
The rounding is automatically done towards zero since its integer division. If there is a remainder, it is discarded.

Reply:
Yeah, but let's imagine we have challenge, where the challenge creator should be charged for 3 coins, and he only has 2 friends. Who should get 3th coin?

Reply:
Just tried to convert to `f64` and divide so that each friend would get `1.5` but `AnsAsset` wont allow this. We could randomly distribute the remainder to one of the friends, lucky prize.

Reply:
Or burn it :rofl:

Reply:
So I originally opted for random distribute. But then didn't want to bring in `rand` as it would bloat the contract. Attempted to write something in raw rust, but it would be computationally expensive and not *that* random. 

So instead went with this, the remainder will just stay in the admin's account. And added a 
```rust
.add_attribute("remainder is", reaminder.to_string()))
```
In the response